network: add virtual wlan interface support

yuwata opened the pull request:
Closes #18977.

poettering commented on man/systemd.link.xml:
```diff
@@ -1062,6 +1062,51 @@
   </refsect1>
 
+  <refsect1>
+    <title>[VirtualWLANInterface] Section Options</title>
```
Hmm, we usually don't suffix the sections in .link files with "Interface" (and not with "Link" either), given they are already located in per-*link* files anyway it's hence redundant.

More importantly though: this is weird in .link? We usually create neteowkr devices via .netdev files, not via .link. Hence we depart from that logic? i.e. .link is for configuring device properties, .netdev is for *creating* net devices, and .network for configuring network properties. This one appears clearly a case for .netdev to me

yuwata replied:
Oh. That makes sense for me. I will move this to .netdev. Thanks.

poettering commented on man/systemd.netdev.xml:
```diff
@@ -192,6 +192,9 @@
           <row><entry><varname>ipoib</varname></entry>
           <entry>An IP over Infiniband subinterface.</entry></row>
+
+          <row><entry><varname>wlan</varname></entry>
```
maybe `virtual-wifi`? or `wifi-virtual`?

poettering commented on man/systemd.netdev.xml:
```diff
@@ -2179,6 +2182,42 @@
   </refsect1>
 
+  <refsect1>
+    <title>[WLAN] Section Options</title>
```
As above, maybe name this [VirtualWifi] or [WifiVirtual]

poettering commented on man/systemd.netdev.xml:
```diff
@@ -2179,6 +2182,42 @@
+      <varlistentry>
+        <term><varname>Phy=</varname></term>
+        <listitem>
+          <para>Specifies the name or index of the WLAN physical layer. This option is mandatory.
```
physical layer *device*.

(Maybe also mention that `iw phy` lists these.)

poettering reviewed (commented):
looks good. just some nitpicking over naming.

yuwata commented:
@poettering Thank you for the review.
- Renamed `[WLAN]` -> `[VirtualWLAN]`. When I added SSID= setting, a person said that WiFi is a trademark (though it is non-profit), and the kernel uses wlan instead of wifi. So, I also use 'wlan' in our manual. Here, I follow that for consistency.
- Hopefully fixed CI failures.

PTAL.

poettering commented on man/systemd.netdev.xml:
```diff
@@ -2179,6 +2182,43 @@
+        <term><varname>Phy=</varname></term>
+        <listitem>
+          <para>Specifies the name or index of the WLAN physical layer device. The list of the physical
+          layer devices exist os the host can be obtained by <command>iw phy</command> command. This
```
exist os → that exist on

keszybz commented on man/systemd.netdev.xml:
```diff
@@ -192,6 +192,9 @@
           <entry>An IP over Infiniband subinterface.</entry></row>
+
+          <row><entry><varname>virtual-wlan</varname></entry>
+          <entry>A virtual wlan interface.</entry></row>
```
Please say `WLAN (local wireless network)` for clarity.

yuwata commented:
@poettering Thanks. The above point is addressed. Upgrading the green label.

keszybz commented on man/systemd.netdev.xml:
```diff
@@ -2179,6 +2182,43 @@
+        <term><varname>WDS=</varname></term>
+        <listitem>
+          <para>Enables the Wireless Distribution System (WDS) mode on the interface. The mode is also
+          known as the 4 address mode. Takes a boolean value. Defaults to unset, and the kernel's
```
Maybe add quotes around "4 address mode" ?

keszybz commented on man/systemd.netdev.xml:
```diff
@@ -2179,6 +2182,43 @@
 
+  <refsect1>
+    <title>[VirtualWLAN] Section Options</title>
+    <para>The [VirtualWLAN] section only applies for virtual WLAN interface, and accepts the following
```
"applies for …" → "applies to virtual WLAN interfaces"

keszybz commented on man/systemd.netdev.xml:
```diff
@@ -2179,6 +2182,43 @@
+      <varlistentry>
+        <term><varname>Phy=</varname></term>
+        <listitem>
+          <para>Specifies the name or index of the WLAN physical layer device. The list of the physical
```
I think the "Phy=" name is not very good. Can we call this "PhysicalDevice="? (It can be called "phy" internally in the code, that is OK.)

Also, "Specifies the name or index of the physical WLAN device." ?

keszybz commented:
Some more late comments, sorry.

gdamjan commented:
So, is the matching of the physical device always be just the numbered physical device (like `phy0`)?

That seems a bit nondeterministic.

Maybe it can be optionally matched on the PCI device or some other attribute?

yuwata commented:
@gdamjan Could you open a new RFE? Otherwise, we easily forget the point. Thank you.

zhangyoufu commented:
The virtual wlan interface does not trigger udev rules like `SUBSYSTEM=="net", KERNEL!="lo", TAG+="systemd", ENV{SYSTEMD_ALIAS}+="/sys/subsystem/net/devices/$name"`. Is there anything wrong?

```
# udevadm info /sys/class/net/phy0-ap0 | cat
P: /devices/pci0000:00/0000:00:02.1/0000:02:00.0/0000:03:08.0/0000:06:00.0/0000:07:05.0/0000:0a:00.0/net/phy0-ap0
M: phy0-ap0
R: 0
J: n19
U: net
T: wlan
I: 19
E: DEVPATH=/devices/pci0000:00/0000:00:02.1/0000:02:00.0/0000:03:08.0/0000:06:00.0/0000:07:05.0/0000:0a:00.0/net/phy0-ap0
E: SUBSYSTEM=net
E: DEVTYPE=wlan
E: INTERFACE=phy0-ap0
E: IFINDEX=19
```

yuwata commented:
> The virtual wlan interface does not trigger udev rules like `SUBSYSTEM=="net", KERNEL!="lo", TAG+="systemd", ENV{SYSTEMD_ALIAS}+="/sys/subsystem/net/devices/$name"`. Is there anything wrong?
> 
> ```
> # udevadm info /sys/class/net/phy0-ap0 | cat
> P: /devices/pci0000:00/0000:00:02.1/0000:02:00.0/0000:03:08.0/0000:06:00.0/0000:07:05.0/0000:0a:00.0/net/phy0-ap0
> M: phy0-ap0
> R: 0
> J: n19
> U: net
> T: wlan
> I: 19
> E: DEVPATH=/devices/pci0000:00/0000:00:02.1/0000:02:00.0/0000:03:08.0/0000:06:00.0/0000:07:05.0/0000:0a:00.0/net/phy0-ap0
> E: SUBSYSTEM=net
> E: DEVTYPE=wlan
> E: INTERFACE=phy0-ap0
> E: IFINDEX=19
> ```

Maybe `NAME=` is unset?
Try to generate debugging logs for the interface by creating a rules file something like the following.
```
# /etc/udev/rules.d/00-debug-net.rules
SUBSYSTEM=="net", OPTIONS="log_level=debug"
```
Also, `udevadm monitor` and `udevadm test` should be your friend. Good luck.

zhangyoufu commented:
@yuwata Thank you for the debugging hints. There is a bug in my systemd .link file and everything works after fixing it.

---

Current implementation have the following pitfalls:
- Hard-coded phy interface index/name. This is not ideal. Due to hot-plug USB Wi-Fi adapter (or faulty PCIe link maybe), the phy interface index subjects to change.
- Cannot assign vif mac address. Sysadmins may want precise control on BSSID. (I am not sure whether hostapd can override nic hwaddr)

I came up with the following udev rules which finds appropriate phy interface by persistent MAC address and spawn vif with MAC address +1/+2.

```
#/etc/udev/rules.d/99-wlan.rules

## remove the default STA interface (aka. auto vif)
ACTION=="add", SUBSYSTEM=="net", ENV{DEVTYPE}=="wlan", ENV{INTERFACE}=="wlan*", PROGRAM=="/usr/bin/iw dev $env{INTERFACE} del"

## create AP interfaces
ACTION=="add", SUBSYSTEM=="ieee80211", ATTR{macaddress}=="00:53:00:00:00:00", RUN+="/usr/bin/iw phy $name interface add ap0 type __ap addr 02:53:00:00:00:01", RUN+="/usr/bin/iw phy $name interface add ap1 type __ap addr 02:53:00:00:00:02"
```
